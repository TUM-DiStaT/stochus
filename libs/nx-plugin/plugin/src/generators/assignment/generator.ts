import {
  UnitTestRunner,
  libraryGenerator as angularLibraryGenerator,
} from '@nx/angular/generators'
import { Tree, formatFiles, generateFiles } from '@nx/devkit'
import { Linter } from '@nx/eslint'
import { libraryGenerator } from '@nx/js'
import { tsquery } from '@phenomnomnominal/tsquery'
import { camelCase, kebabCase, upperFirst } from 'lodash'
import * as path from 'path'
import {
  backendAssignmentsServicePath,
  frontendAssignmentsServicePath,
} from './file-paths'
import { AssignmentGeneratorSchema } from './schema'

function addToFrontendService(
  tree: Tree,
  camelCasedName: string,
  options: AssignmentGeneratorSchema,
) {
  const frontendAssignmentService =
    tree.read(frontendAssignmentsServicePath)?.toString() ?? ''
  const withImport = tsquery.replace(
    frontendAssignmentService,
    // last-child doesn't seem to work and the order will be fixed by prettier
    'ImportDeclaration:first-child',
    (node) => {
      return (
        node.getText() +
        '\n' +
        `import { ${camelCasedName}AssignmentForFrontend } from '@stochus/assignments/${options.name}/frontend'`
      )
    },
  )
  const withAddedAssignment = tsquery.replace(
    withImport,
    'ClassDeclaration:has(Identifier[name="AssignmentsService"]) > PropertyDeclaration:has(Identifier[name="assignments"]) > ArrayLiteralExpression AsExpression:last-child',
    (node) => {
      return (
        node.getText() +
        `,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ${camelCasedName}AssignmentForFrontend as any`
      )
    },
  )
  tree.write(frontendAssignmentsServicePath, withAddedAssignment)
}

function addToBackendService(
  tree: Tree,
  camelCasedName: string,
  options: AssignmentGeneratorSchema,
) {
  const backendAssignmentService =
    tree.read(backendAssignmentsServicePath)?.toString() ?? ''
  const withImport = tsquery.replace(
    backendAssignmentService,
    // last-child doesn't seem to work and the order will be fixed by prettier
    'ImportDeclaration:first-child',
    (node) => {
      return (
        node.getText() +
        '\n' +
        `import { ${camelCasedName}Assignment } from '@stochus/assignments/${options.name}/shared'`
      )
    },
  )
  const withAddedAssignment = tsquery.replace(
    withImport,
    'ClassDeclaration:has(Identifier[name="AssignmentsCoreBackendService"]) > PropertyDeclaration:has(Identifier[name="assignments"]) > ArrayLiteralExpression Identifier:nth-last-child(2)',
    (node) => {
      return (
        node.getText() +
        `,
        ${camelCasedName}Assignment`
      )
    },
  )
  tree.write(backendAssignmentsServicePath, withAddedAssignment)
}

export async function assignmentGenerator(
  tree: Tree,
  options: AssignmentGeneratorSchema,
) {
  const projectRoot = `libs/assignments/${options.name}`
  const camelCasedName = upperFirst(camelCase(options.name))

  await libraryGenerator(tree, {
    name: `${options.name}-assignment-shared`,
    tags: 'scope:shared',
    importPath: `@stochus/assignments/${options.name}/shared`,
    minimal: true,
    publishable: false,
    linter: 'eslint',
    strict: true,
    unitTestRunner: 'jest',
    buildable: false,
    directory: projectRoot + '/shared',
    projectNameAndRootFormat: 'as-provided',
    config: 'project',
    compiler: 'tsc',
  })
  await angularLibraryGenerator(tree, {
    name: `${options.name}-assignment-frontend`,
    tags: 'scope:frontend',
    buildable: false,
    linter: Linter.EsLint,
    strict: true,
    publishable: false,
    directory: projectRoot + '/frontend',
    importPath: `@stochus/assignments/${options.name}/frontend`,
    unitTestRunner: UnitTestRunner.Jest,
    projectNameAndRootFormat: 'as-provided',
    prefix: 'stochus',
    standalone: false,
    skipModule: true,
    skipSelector: true,
    standaloneConfig: true,
  })

  // Remove certain generated files that won't be needed
  const sharedFilesToDelete = [
    `src/lib/${options.name}-assignment-shared.ts`,
    `src/lib/${options.name}-assignment-shared.spec.ts`,
  ]
  for (const file of sharedFilesToDelete) {
    tree.delete(`${projectRoot}/shared/${file}`)
  }

  addToFrontendService(tree, camelCasedName, options)
  addToBackendService(tree, camelCasedName, options)

  generateFiles(tree, path.join(__dirname, 'files'), projectRoot, {
    ...options,
    name: kebabCase(options.name),
    camelCasedName,
  })
  await formatFiles(tree)
}

export default assignmentGenerator
