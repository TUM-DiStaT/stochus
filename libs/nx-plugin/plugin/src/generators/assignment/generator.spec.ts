import { Tree, readProjectConfiguration, updateJson } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import { tsquery } from '@phenomnomnominal/tsquery'
import * as fs from 'fs/promises'
import { camelCase, upperFirst } from 'lodash'
import * as path from 'path'
import {
  backendAssignmentsServicePath,
  frontendAssignmentsServicePath,
} from './file-paths'
import { assignmentGenerator } from './generator'
import { AssignmentGeneratorSchema } from './schema'

const getActualFileContentFromWorkspaceRootPath = async (filePath: string) =>
  await fs.readFile(
    path.join(__dirname, '../../../../../../', filePath),
    'utf-8',
  )

const getActualPrettierConfig = async () => {
  return JSON.parse(
    await getActualFileContentFromWorkspaceRootPath('.prettierrc'),
  )
}

describe('assignment generator', () => {
  let tree: Tree
  const options: AssignmentGeneratorSchema = { name: 'do-a-backflip' }
  const assignmentBaseDir = `libs/assignments/${options.name}`
  let actualPrettierConfig: object

  beforeAll(async () => {
    actualPrettierConfig = await getActualPrettierConfig()
  })

  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace()
    expect(tree.exists('.prettierrc')).toBe(true)
    updateJson(tree, '.prettierrc', (json) => ({
      ...json,
      ...actualPrettierConfig,
    }))

    // Tell the library generator where to create the new project
    process.env.INIT_CWD = tree.root
    jest.spyOn(process, 'cwd').mockReturnValue(tree.root)

    // Avoid warning messages from angular generator that no .gitignore can be found
    tree.write('.gitignore', '')

    // Add current assignment services
    const frontendAssignmentsServiceContent =
      await getActualFileContentFromWorkspaceRootPath(
        frontendAssignmentsServicePath,
      )
    tree.write(
      frontendAssignmentsServicePath,
      frontendAssignmentsServiceContent,
    )
    const backendAssignmentsServiceContent =
      await getActualFileContentFromWorkspaceRootPath(
        backendAssignmentsServicePath,
      )
    tree.write(backendAssignmentsServicePath, backendAssignmentsServiceContent)
  })

  it('should create shared project', async () => {
    await assignmentGenerator(tree, options)
    const config = readProjectConfiguration(
      tree,
      `${options.name}-assignment-shared`,
    )
    expect(config).toBeDefined()
    expect(config.tags).toContain('scope:shared')
    expect(config.projectType).toBe('library')
  })

  it('should create frontend project', async () => {
    await assignmentGenerator(tree, options)
    const config = readProjectConfiguration(
      tree,
      `${options.name}-assignment-frontend`,
    )
    expect(config).toBeDefined()
    expect(config.tags).toContain('scope:frontend')
    expect(config.projectType).toBe('library')
  })

  it.each([
    `${assignmentBaseDir}/shared/src/lib/${options.name}-assignment-shared.ts`,
    `${assignmentBaseDir}/shared/src/lib/${options.name}-assignment-shared.spec.ts`,
    `${assignmentBaseDir}/frontend/src/lib/${options.name}-assignment-frontend.component.ts`,
    `${assignmentBaseDir}/frontend/src/lib/${options.name}-assignment-frontend.component.spec.ts`,
    `${assignmentBaseDir}/frontend/src/lib/${options.name}-assignment-frontend.component.html`,
    `${assignmentBaseDir}/frontend/src/lib/${options.name}-assignment-frontend.component.css`,
  ])('should delete the unneeded file %p', async (filePath) => {
    await assignmentGenerator(tree, options)
    expect(tree.exists(filePath)).toBe(false)
  })

  it.each([
    `${assignmentBaseDir}/shared/src/index.ts`,
    `${assignmentBaseDir}/shared/src/lib/${options.name}-assignment.ts`,
    `${assignmentBaseDir}/shared/src/lib/${options.name}-assignment.spec.ts`,

    `${assignmentBaseDir}/frontend/src/index.ts`,
    `${assignmentBaseDir}/frontend/src/lib/${options.name}-assignment-for-frontend.ts`,
    `${assignmentBaseDir}/frontend/src/lib/assignment-process/${options.name}-assignment-process.component.html`,
    `${assignmentBaseDir}/frontend/src/lib/assignment-process/${options.name}-assignment-process.component.ts`,
    `${assignmentBaseDir}/frontend/src/lib/assignment-process/${options.name}-assignment-process.component.spec.ts`,
    `${assignmentBaseDir}/frontend/src/lib/config-form/${options.name}-config-form.component.html`,
    `${assignmentBaseDir}/frontend/src/lib/config-form/${options.name}-config-form.component.ts`,
    `${assignmentBaseDir}/frontend/src/lib/config-form/${options.name}-config-form.component.spec.ts`,
    `${assignmentBaseDir}/frontend/src/lib/feedback/${options.name}-feedback.component.html`,
    `${assignmentBaseDir}/frontend/src/lib/feedback/${options.name}-feedback.component.ts`,
    `${assignmentBaseDir}/frontend/src/lib/feedback/${options.name}-feedback.component.spec.ts`,
  ])('should create %p with the correct contents', async (fileName) => {
    await assignmentGenerator(tree, options)
    expect(tree.exists(fileName)).toBe(true)
    expect(tree.read(fileName)?.toString()).toMatchSnapshot()
  })

  it('should add the asssignment to the frontend service', async () => {
    await assignmentGenerator(tree, options)
    const assignmentServiceCode =
      tree.read(frontendAssignmentsServicePath)?.toString() ?? ''
    expect(assignmentServiceCode).not.toBe('')

    const assignmentForFrontendDefinitionName =
      upperFirst(camelCase(options.name)) + 'AssignmentForFrontend'

    // Import should exist
    const importLiteralNodes = tsquery.query(
      assignmentServiceCode,
      `ImportDeclaration:has(Identifier[name="${assignmentForFrontendDefinitionName}"]):has(StringLiteral[value="@stochus/assignments/${options.name}/frontend"])`,
    )
    expect(importLiteralNodes).toHaveLength(1)

    // Assignment should be added to the array
    const assignmentsArrayLiteralNodes = tsquery.query(
      assignmentServiceCode,
      'ClassDeclaration:has(Identifier[name="AssignmentsService"]) ' +
        '> PropertyDeclaration:has(Identifier[name="assignments"]) ' +
        `> ArrayLiteralExpression`,
    )
    expect(assignmentsArrayLiteralNodes).toHaveLength(1)
    const assignmentsArrayLiteral = assignmentsArrayLiteralNodes[0]
    expect(assignmentsArrayLiteral.getFullText()).toMatch(
      assignmentForFrontendDefinitionName + ' as any',
    )
    // It should be added with the eslint-disable-next-line comment
    expect(assignmentsArrayLiteral.getFullText()).toMatch(
      /\/\/ eslint-disable-next-line @typescript-eslint\/no-explicit-any\n\s*DoABackflipAssignmentForFrontend as any/,
    )

    // This final assertion is just meant for manual review of the final file.
    // If it ever fails, the assertions above *should* be enough to ensure functionality.
    // Still, take a second to look at the changes and see if this code would still work.
    expect(assignmentServiceCode).toMatchSnapshot()
  })

  it('should add the asssignment to the backend service', async () => {
    await assignmentGenerator(tree, options)
    const assignmentServiceCode =
      tree.read(backendAssignmentsServicePath)?.toString() ?? ''
    expect(assignmentServiceCode).not.toBe('')

    const assignmentDefinitionName =
      upperFirst(camelCase(options.name)) + 'Assignment'

    // Import should exist
    const importLiteralNodes = tsquery.query(
      assignmentServiceCode,
      `ImportDeclaration:has(Identifier[name="${assignmentDefinitionName}"]):has(StringLiteral[value="@stochus/assignments/${options.name}/shared"])`,
    )
    expect(importLiteralNodes).toHaveLength(1)

    // Assignment should be added to the array
    const assignmentsArrayLiteralNodes = tsquery.query(
      assignmentServiceCode,
      'ClassDeclaration:has(Identifier[name="AssignmentsCoreBackendService"]) > PropertyDeclaration:has(Identifier[name="assignments"]) > ArrayLiteralExpression Identifier:nth-last-child(2)',
    )
    expect(assignmentsArrayLiteralNodes).toHaveLength(1)
    const assignmentsArrayLiteral = assignmentsArrayLiteralNodes[0]
    expect(assignmentsArrayLiteral.getFullText()).toMatch(
      assignmentDefinitionName,
    )

    // This final assertion is just meant for manual review of the final file.
    // If it ever fails, the assertions above *should* be enough to ensure functionality.
    // Still, take a second to look at the changes and see if this code would still work.
    expect(assignmentServiceCode).toMatchSnapshot()
  })
})
