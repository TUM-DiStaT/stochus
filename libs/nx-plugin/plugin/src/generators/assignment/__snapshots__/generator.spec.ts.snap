// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assignment generator should create "libs/assignments/do-a-backflip/frontend/src/index.ts" with the correct contents 1`] = `
"export * from './lib/do-a-backflip-assignment-for-frontend';
"
`;

exports[`assignment generator should create "libs/assignments/do-a-backflip/shared/src/index.ts" with the correct contents 1`] = `
"export * from './lib/do-a-backflip-assignment';
"
`;
